File: packages/wasmvm/wasmlib/src/wasmtypes/scchainid.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_CHAIN_ID_LENGTH: usize = 32;

#[derive(PartialEq, Copy, Clone)]
pub struct ScChainID {
    id: [u8; SC_CHAIN_ID_LENGTH],
}

impl ScChainID {
    pub fn address(&self) -> ScAddress {
        let buf = [SC_ADDRESS_ALIAS];
        address_from_bytes(&[&buf[..], &self.id[..]].concat())
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        chain_id_to_bytes(self)
    }

    pub fn to_string(&self) -> String {
        chain_id_to_string(self)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn chain_id_decode(dec: &mut WasmDecoder) -> ScChainID {
    chain_id_from_bytes_unchecked(&dec.fixed_bytes(SC_CHAIN_ID_LENGTH))
}

pub fn chain_id_encode(enc: &mut WasmEncoder, value: &ScChainID) {
    enc.fixed_bytes(&value.id, SC_CHAIN_ID_LENGTH);
}

pub fn chain_id_from_bytes(buf: &[u8]) -> ScChainID {
    if buf.len() == 0 {
        return ScChainID {
            id: [0; SC_CHAIN_ID_LENGTH],
        };
    }
    if buf.len() != SC_CHAIN_ID_LENGTH {
        panic("invalid ChainID length");
    }
    ScChainID {
        id: buf.try_into().expect("WTF?"),
    }
}

pub fn chain_id_to_bytes(value: &ScChainID) -> Vec<u8> {
    value.id.to_vec()
}

pub fn chain_id_from_string(value: &str) -> ScChainID {
    let addr = address_from_string(value);
    if addr.id[0] != SC_ADDRESS_ALIAS {
        panic("invalid ChainID address type");
    }
    chain_id_from_bytes(&addr.id[1..])
}

pub fn chain_id_to_string(value: &ScChainID) -> String {
    address_to_string(&value.address())
}

fn chain_id_from_bytes_unchecked(buf: &[u8]) -> ScChainID {
    ScChainID {
        id: buf.try_into().expect("invalid ChainID length"),
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableChainID {
    proxy: Proxy,
}

impl ScImmutableChainID {
    pub fn new(proxy: Proxy) -> ScImmutableChainID {
        ScImmutableChainID { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        chain_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScChainID {
        chain_id_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScChainID in host container
pub struct ScMutableChainID {
    proxy: Proxy,
}

impl ScMutableChainID {
    pub fn new(proxy: Proxy) -> ScMutableChainID {
        ScMutableChainID { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: &ScChainID) {
        self.proxy.set(&chain_id_to_bytes(&value));
    }

    pub fn to_string(&self) -> String {
        chain_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScChainID {
        chain_id_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scchainid.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from "../sandbox";
import {WasmDecoder, WasmEncoder, zeroes} from "./codec";
import {Proxy} from "./proxy";
import {addressFromBytes, addressFromString, addressToString, ScAddress, ScAddressAlias} from "./scaddress";
import {bytesCompare} from "./scbytes";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScChainIDLength = 32;

export class ScChainID {
    id: Uint8Array = zeroes(ScChainIDLength);

    public address(): ScAddress {
        const buf = new Uint8Array(this.id.length + 1);
        buf[0] = ScAddressAlias as u8;
        buf.set(this.id, 1);
        return addressFromBytes(buf);
    }

    public equals(other: ScChainID): bool {
        return bytesCompare(this.id, other.id) == 0;
    }

    // convert to byte array representation
    public toBytes(): Uint8Array {
        return chainIDToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        return chainIDToString(this)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function chainIDDecode(dec: WasmDecoder): ScChainID {
    return chainIDFromBytesUnchecked(dec.fixedBytes(ScChainIDLength));
}

export function chainIDEncode(enc: WasmEncoder, value: ScChainID): void {
    enc.fixedBytes(value.id, ScChainIDLength);
}

export function chainIDFromBytes(buf: Uint8Array | null): ScChainID {
    if (buf === null || buf.length == 0) {
        return new ScChainID();
    }
    if (buf.length != ScChainIDLength) {
        panic("invalid ChainID length");
    }
    return chainIDFromBytesUnchecked(buf);
}

export function chainIDToBytes(value: ScChainID): Uint8Array {
    return value.id;
}

export function chainIDFromString(value: string): ScChainID {
    const addr = addressFromString(value);
    if (addr.id[0] != ScAddressAlias) {
        panic("invalid ChainID address type");
    }
    return chainIDFromBytes(addr.id.slice(1));
}

export function chainIDToString(value: ScChainID): string {
    return addressToString(value.address());
}

function chainIDFromBytesUnchecked(buf: Uint8Array): ScChainID {
    let o = new ScChainID();
    o.id = buf.slice(0);
    return o;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableChainID {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return chainIDToString(this.value());
    }

    value(): ScChainID {
        return chainIDFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableChainID extends ScImmutableChainID {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScChainID): void {
        this.proxy.set(chainIDToBytes(value));
    }
}
